smart_open/asb.py

```diff
@@ -0,0 +1,400 @@
+            #
+            binary = b''
+        elif size == -1:
+            binary = self._blob.download_blob().readall()
+        return binary
+
+class SeekableBufferedInputBase(io.BufferedIOBase):
+    """Reads bytes from Azure Blob Storage.
+
+    Implements the io.BufferedIOBase interface of the standard library.
+
+    :raises ResourceNotFoundError: Raised when the blob to read from does not exist.
+
+    """
+    def __init__(
+            self,
+            container,
+            blob,
+            buffer_size=DEFAULT_BUFFER_SIZE,
+            line_terminator=_BINARY_NEWLINE,
+            client=None,  # type: azure.storage.blob.BlobServiceClient
+    ):
+        if client is None:
+            client = BlobServiceClient()
+        container_client = client.get_container_client(container)  # type: azure.storage.blob.ContainerClient
+
+        self._blob = container_client.get_blob_client(blob)
+        if self._blob is None:
+            raise ResourceNotFoundError('blob {} not found in {}'.format(blob, container))
+        self._size = self._blob.get_blob_properties().size if self._blob.get_blob_properties().size is not None else 0
+
+        self._raw_reader = _SeekableRawReader(self._blob, self._size)
+        self._current_pos = 0
+        self._current_part_size = buffer_size
+        self._current_part = smart_open.bytebuffer.ByteBuffer(buffer_size)
+        self._eof = False
+        self._line_terminator = line_terminator
+
+        #
+        # This member is part of the io.BufferedIOBase interface.
+        #
+        self.raw = None
+
+    #
+    # Override some methods from io.IOBase.
+    #
+    def close(self):
+        """Flush and close this stream."""
+        logger.debug("close: called")
+        self._blob = None
+        self._raw_reader = None
+
+    def readable(self):
+        """Return True if the stream can be read from."""
+        return True
+
+    def seekable(self):
+        """If False, seek(), tell() and truncate() will raise IOError.
+
+        We offer only seek support, and no truncate support."""
+        return True
+
+    #
+    # io.BufferedIOBase methods.
+    #
+    def detach(self):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def seek(self, offset, whence=START):
+        """Seek to the specified position.
+
+        :param int offset: The offset in bytes.
+        :param int whence: Where the offset is from.
+
+        Returns the position after seeking."""
+        logger.debug('seeking to offset: %r whence: %r', offset, whence)
+        if whence not in _WHENCE_CHOICES:
+            raise ValueError('invalid whence, expected one of %r' % _WHENCE_CHOICES)
+
+        if whence == START:
+            new_position = offset
+        elif whence == CURRENT:
+            new_position = self._current_pos + offset
+        else:
+            new_position = self._size + offset
+        new_position = smart_open.s3.clamp(new_position, 0, self._size)
```

Comment:
Is `clamp` specific to S3 ? Maybe put it in a separate file `helper.py` or `util.py` to have better package imports, cc @Tyrannas @mpenkov, that kind of packaging is related to #443

Reply:
Yes, see the util submodule.